File: C++/time_based_key_value_store.cpp
class TimeMap {
private:
    unordered_map<string, vector<pair<int, string>>> timeMap;

public:
    TimeMap() {
    }
    
    void set(string key, string value, int timestamp) {
        timeMap[key].push_back({timestamp, value});
    }
    
    string get(string key, int timestamp) {
        string res = "";
        if (!timeMap.contains(key)){return res;}

        int l = 0; int r = timeMap[key].size() - 1;
        vector<pair<int, string>>& vals = timeMap[key];

        while (l <= r){
            int mid = l + (r - l) / 2;
            if (vals[mid].first == timestamp){
                res = vals[mid].second;
                break;
            }

            else if (vals[mid].first < timestamp){
                res = vals[mid].second;
                l = mid + 1;
            }

            else{r = mid - 1;}
        }

        return res;
    }
};

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap* obj = new TimeMap();
 * obj->set(key,value,timestamp);
 * string param_2 = obj->get(key,timestamp);
 */

File: python/time_based_key_value_store.py
class TimeMap:

    def __init__(self):
        self.timeMap = defaultdict(list)
        

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.timeMap[key].append([timestamp, value])
        

    def get(self, key: str, timestamp: int) -> str:
        res = ''
        vals = self.timeMap.get(key, [])
        l, r = 0, len(vals) - 1

        while l <= r:
            mid = l + (r - l) // 2
            if vals[mid][0] == timestamp:
                res = vals[mid][1]
                break
            elif vals[mid][0] < timestamp:
                res = vals[mid][1]
                l = mid + 1
            else:
                r = mid -1
        return res
        


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)
